src/FwdState.cc

```diff
@@ -896,12 +907,87 @@ FwdState::noteConnection(HappyConnOpener::Answer &answer)
         if (originWantsEncryptedTraffic && // the "encrypted traffic" part
                 !peer->options.originserver && // the "through a proxy" part
                 !peer->secure.encryptTransport) // the "exclude HTTPS proxies" part
-            return advanceDestination("establish tunnel through proxy", answer.conn, [this,&answer] {
-            establishTunnelThruProxy(answer.conn);
+            return advanceDestination("establish tunnel through proxy", conn, [this,&conn] {
+            establishTunnelThruProxy(conn);
         });
     }
 
-    secureConnectionToPeerIfNeeded(answer.conn);
+    secureConnectionToPeerIfNeeded(conn);
+}
+
+/// syncs proxyProtocolHeader with the current request forwarding attempt
+void
+FwdState::resetProxyProtocolHeader()
+{
+    proxyProtocolHeader.reset();
+
+    if (!Config.outgoingProxyProtocolHttp)
+        return;
+
+    if (const auto &aclList = Config.outgoingProxyProtocolHttp->aclList) {
+        ACLFilledChecklist ch(aclList, request);
+        ch.al = al;
+        ch.syncAle(request, nullptr);
+        if (!ch.fastCheck().allowed())
+            return;
+    }
+
+    static const SBuf v2("2.0");
+    ProxyProtocol::Header header(v2, request->masterXaction->initiator.internalClient() ? ProxyProtocol::Two::cmdLocal : ProxyProtocol::Two::cmdProxy);
+    Config.outgoingProxyProtocolHttp->fill(header, al);
+
+    BinaryPacker packer;
+    header.pack(packer);
+    proxyProtocolHeader = packer.packed();
+}
+
+void
+FwdState::sendProxyProtoHeaderIfNeeded(const Comm::ConnectionPointer &conn)
+{
+    if (!proxyProtocolHeader) {
+        tunnelIfNeeded(conn);
+        return;
+    }
+
```

Comment:
Please use advanceDestination() for the code below, just like other existing advanceDestination() callers are doing for the corresponding steps that "advance" destination/connection towards dispatch()-ready state OR discuss why we should make an exception for sending PROXY protocol header.

Reply:
Done (1f52144).